truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -1136,11 +1196,176 @@ public Object divSpecialNormal(VirtualFrame frame, RubyBasicObject a, RubyBasicO
+            return createArray(createBigDecimal(frame, result[0]), createBigDecimal(frame, result[1]));
+        }
+
+        @Specialization(guards = {
+                "isNormal(a)",
+                "isNormalRubyBigDecimal(b)",
+                "isNormalZero(a)",
+                "!isNormalZero(b)" })
+        public Object divmodZeroDividend(VirtualFrame frame, RubyBasicObject a, RubyBasicObject b) {
+            return createArray(
+                    createBigDecimal(frame, BigDecimal.ZERO),
+                    createBigDecimal(frame, BigDecimal.ZERO));
+        }
+
+        @Specialization(guards = {
+                "isNormal(a)",
+                "isNormalRubyBigDecimal(b)",
+                "isNormalZero(b)" })
+        public Object divmodZeroDivisor(VirtualFrame frame, RubyBasicObject a, RubyBasicObject b) {
+            throw new RaiseException(getContext().getCoreLibrary().zeroDivisionError(this));
+        }
+
+        @Specialization(guards = {
+                "isRubyBigDecimal(b)",
+                "!isNormal(a) || !isNormal(b)" })
+        public Object divmodSpecial(VirtualFrame frame, RubyBasicObject a, RubyBasicObject b) {
+            final Type aType = getBigDecimalType(a);
+            final Type bType = getBigDecimalType(b);
+
+            if (aType == Type.NAN || bType == Type.NAN) {
+                return createArray(createBigDecimal(frame, Type.NAN), createBigDecimal(frame, Type.NAN));
+            }
+
+            if (bType == Type.NEGATIVE_ZERO || (bType == Type.NORMAL && isNormalZero(b))) {
+                throw new RaiseException(getContext().getCoreLibrary().zeroDivisionError(this));
+            }
+
+            if (aType == Type.NEGATIVE_ZERO || (aType == Type.NORMAL && isNormalZero(a))) {
+                return createArray(
+                        createBigDecimal(frame, BigDecimal.ZERO),
+                        createBigDecimal(frame, BigDecimal.ZERO));
+            }
+
+            if (aType == Type.POSITIVE_INFINITY || aType == Type.NEGATIVE_INFINITY) {
+                setupSignCall();
+                setupLimitIntegerCast();
+
+                final int signA = aType == Type.POSITIVE_INFINITY ? 1 : -1;
+                final int signB = Integer.signum(signIntegerCast.executeInteger(frame, signCall.call(frame, b, "sign", null)));
+                final int sign = signA * signB; // is between -1 and 1, 0 when nan
+
+                final Type type = new Type[]{ Type.NEGATIVE_INFINITY, Type.NAN, Type.POSITIVE_INFINITY }[sign + 1];
+
+                return createArray(
+                        createBigDecimal(frame, type),
+                        createBigDecimal(frame, Type.NAN));
+            }
+
+            if (bType == Type.POSITIVE_INFINITY || bType == Type.NEGATIVE_INFINITY) {
+                return createArray(createBigDecimal(frame, BigDecimal.ZERO), createBigDecimal(frame, a));
+            }
+
+            throw new UnsupportedOperationException();
```

Comment:
While we have a lot of `UnsupportedOperationException`s without a message, I think we should really try to use a descriptive error message in all future work.